watch no longer triggers update; adds SignalWatcher.effect

sorvell opened the pull request:
Fixes https://github.com/lit/lit/issues/4860.

* `watch` no longer triggers a reactive update on a `SignalWatcher` element
* `watch` can be used without `SignalWatcher`
* adds `SignalWatcher.effect` which is called by default after rendering updates but can optionally include an options `{beforeUpdate: true}` to process the effect before any rendering updates.

justinfagnani commented on packages/labs/signals/src/lib/signal-watcher.ts:
```diff
@@ -4,23 +4,79 @@
+ * change. If an options object is provided, the `element` property can be used
+ * to specify the element to associate the effect with. The `beforeUpdate`
+ * property can be used to specify that the effect should run before the element
+ * updates. The `manualDispose` property can be used to specify that the effect
```
What's the use case for `manualDispose`?

sorvell replied:
Clarified docs

sorvell replied:
Added test for `manualDispose`

justinfagnani commented on packages/labs/signals/src/lib/signal-watcher.ts:
```diff
@@ -4,23 +4,79 @@
+ *
+ * @param callback
+ * @param options {element, beforeUpdate, manualDispose}
+ * @returns
```
remove `@returns` or add the type?

sorvell replied:
Done

justinfagnani commented on packages/labs/signals/src/lib/signal-watcher.ts:
```diff
@@ -4,23 +4,79 @@
+}
+
+export interface ElementEffectOptions extends EffectOptions {
+  element?: SignalWatcher & SignalWatcherApi;
```
JSDocs here?

sorvell replied:
Removed, no longer needed.

justinfagnani commented on packages/labs/signals/src/lib/signal-watcher.ts:
```diff
@@ -4,23 +4,79 @@
+};
+
+interface SignalWatcherApi {
+  _effect(fn: () => void, options?: EffectOptions): () => void;
```
Should this be marked `/** @internal */`?

sorvell replied:
Done

sorvell replied:
@justinfagnani if we only run effect for element, should we just expose it directly here? Maybe as `signalEffect` to avoid taking the generic name?

justinfagnani replied:
Yeah, I think `this.signalEffect()` is ok, thought maybe redundant since "effect" is strongly associated with signals in this world already. `this.updateEffect()` might be ok too? Or does "update" sound like a verb in that usage?

I do think an instance method is fine for this though. As opposed to `effect(this, () => ...)`.

sorvell replied:
changed to `updateEffect`

justinfagnani commented on packages/labs/signals/src/lib/signal-watcher.ts:
```diff
@@ -48,24 +101,61 @@ const elementForWatcher = new WeakMap<
+        );
+      beforeEffects.forEach((signal) => signal.get());
+      this.__performUpdateSignal?.get();
+      pendingWatches!.forEach((signal) => signal.get());
```
Should the pending watches only be run if the full update didn't run? Otherwise those bindings will run twice?

sorvell replied:
No, the get won't provoke it running twice if the value hasn't changed, right?

justinfagnani replied:
ah, correct

justinfagnani commented on packages/labs/signals/src/lib/signal-watcher.ts:
```diff
@@ -87,25 +177,70 @@ export function SignalWatcher<T extends ReactiveElementConstructor>(
-      this.__watcher.unwatch(this.__performUpdateSignal!);
+      // We unwatch all signals that are not manually disposed, so that we don't
+      // keep the element alive by holding references to it.
+      this._watcher.unwatch(
```
Do we do this manually because of manualDispose, or is there also no bulk way to dispose of the whole watcher?

sorvell commented on packages/labs/signals/src/lib/signal-watcher.ts:
```diff
@@ -4,23 +4,79 @@
+}
+
+interface EffectOptions {
+  beforeUpdate?: boolean;
```
Happy to, but not sure where it goes. Can you point it out in the code?

justinfagnani replied:
You just put them on the option property:

```ts
interface EffectOptions {
  /**
   * Whether to run the effect before the element's update. Defaults to `false`.
   */
  beforeUpdate?: boolean;

  /** ... */
  manualDispose?: boolean;
}
```

sorvell replied:
Done

justinfagnani commented on packages/labs/signals/src/lib/signal-watcher.ts:
```diff
@@ -4,23 +4,60 @@
+  manualDispose?: boolean;
+}
+
+export interface ElementEffectOptions extends EffectOptions {
```
Is this interface used anywhere?

sorvell replied:
Removed, no longer needed.

justinfagnani reviewed (approved):
Awesome!

Two small comments left

Westbrook commented:
Getting into some finer testing of this change as we await it landing and wondered which of these is "correct".

In https://lit.dev/playground/#gist=5cfb3e893dbc5412ed1468483be5fa80 we get the render lifecycle on all three buttons, but only the Signal bindings on the first and last (signal powered) element.

In https://lit.dev/playground/#gist=3b14cb4a58de364ec7686f230a88435d we do not get the render lifecycle in either of the Signal bindings.

Maybe there's changes in this PR before I lifted the code, but wanted to confirm as we're working out how we want to consumer these capabilities in my team right now.

sorvell commented:
Forwarded from [discord](https://discord.com/channels/1012791295170859069/1068571268607127643/1435397282597175390)

* This should be removed: https://lit.dev/docs/data/signals/#:~:text=For%20coordination%20purposes,template%20are%20skipped
* We changed this since signals are not part of the element's reactive properties so if they triggered an update it would be an update without any visible changes to properties.
* We've added `updateEffect` which lets you use signal values when signals change. This method is coordinated with the rendering impact that any of those signals have on the element and can be either before or after them, based on options e.g. `updateEffect(() => console.log(mySIgnal.get()), {beforeUpdate: true})`
